src/coreclr/jit/lowerxarch.cpp

```diff
@@ -6568,27 +6558,43 @@ GenTree* Lowering::PreferredRegOptionalOperand(GenTree* tree)
     //
     // f) If neither of them are local vars (i.e. tree temps), prefer to
     // mark op1 as reg optional for the same reason as mentioned in (d) above.
-    if (op1->OperGet() == GT_LCL_VAR && op2->OperGet() == GT_LCL_VAR)
-    {
-        LclVarDsc* v1 = comp->lvaGetDesc(op1->AsLclVarCommon());
-        LclVarDsc* v2 = comp->lvaGetDesc(op2->AsLclVarCommon());
 
-        bool v1IsRegCandidate = !v1->lvDoNotEnregister;
-        bool v2IsRegCandidate = !v2->lvDoNotEnregister;
-        if (v1IsRegCandidate && v2IsRegCandidate)
+    assert(!op1->IsRegOptional());
+    assert(!op2->IsRegOptional());
+
+    // We default to op1, as op2 is likely to have the shorter lifetime.
+    GenTree* preferredOp = op1;
+
+    if (op1->OperIs(GT_LCL_VAR))
+    {
+        if (op2->OperIs(GT_LCL_VAR))
         {
-            // Both are enregisterable locals.  The one with lower weight is less likely
-            // to get a register and hence beneficial to mark the one with lower
-            // weight as reg optional.
-            // If either is not tracked, it may be that it was introduced after liveness
-            // was run, in which case we will always prefer op1 (should we use raw refcnt??).
-            if (v1->lvTracked && v2->lvTracked && (v1->lvRefCntWtd() >= v2->lvRefCntWtd()))
+            LclVarDsc* v1 = comp->lvaGetDesc(op1->AsLclVarCommon());
+            LclVarDsc* v2 = comp->lvaGetDesc(op2->AsLclVarCommon());
+
+            bool v1IsRegCandidate = !v1->lvDoNotEnregister;
+            bool v2IsRegCandidate = !v2->lvDoNotEnregister;
+
+            if (v1IsRegCandidate && v2IsRegCandidate)
             {
-                preferredOp = op2;
+                // Both are enregisterable locals.  The one with lower weight is less likely
+                // to get a register and hence beneficial to mark the one with lower
+                // weight as reg optional.
+                //
+                // If either is not tracked, it may be that it was introduced after liveness
+                // was run, in which case we will always prefer op1 (should we use raw refcnt??).
+
+                if (v1->lvTracked && v2->lvTracked)
+                {
+                    if (v1->lvRefCntWtd() > v2->lvRefCntWtd())
```

Comment:
Everything except for this one line is consistent with the old code, just cleaned up slightly to reduce the checks.

This one line changed from `(v1->lvRefCntWtd() >= v2->lvRefCntWtd())` to `(v1->lvRefCntWtd() > v2->lvRefCntWtd())` because it mismatched the actual commentary given above, which is that we want to prefer the one with the lower weight and default back to `op1` if all things are equal due to op1 likely having a shorter lifetime and being the preferred option to spill if op2 needs a register

Reply:
This caused some regressions around `INS_cmp`, so I reverted it.